### PR TITLE
[SPARK-5946][Streaming] Add Python API for direct Kafka stream

### PR DESCRIPTION
**Pull request by jerryshao**
Currently only added `createDirectStream` API, I'm not sure if `createRDD` is also needed, since some Java object needs to be wrapped in Python. Please help to review, thanks a lot.

### CODE REVIEW

**Comment by davies** [comment]
@tdas What's the difference between `createStream()` and `createDirectStream()` ? It's possible to combine them together?

**Comment by tdas** [comment]
@davies It is a the new experimental API for Kafka, which does not use Receivers at all, and rather treats Kafka as a file system and topics as files. It finds new data in topic pretty much like your patch to support appends in file stream.

**Comment by davies** [comment]
@tdas Cool, thanks. Do it mean that it does NOT need WAL to have high durability?

This patch looks good to me, just need more docs (somewhere) to tell the difference between these two APIs.

File: examples/src/main/python/streaming/direct_kafka_wordcount.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+if __name__ == "__main__":
+    if len(sys.argv) != 3:
+        print >> sys.stderr, "Usage: direct_kafka_wordcount.py <broker_list> <topic>"
+        exit(-1)
+
+    sc = SparkContext(appName="PythonStreamingDirectKafkaWordCount")
+    ssc = StreamingContext(sc, 2)
+
+    brokers, topic = sys.argv[1:]
+    kvs = KafkaUtils.createDirectStream(ssc, [topic], {"metadata.broker.list": brokers})
```
Comment (davies) [inline examples/src/main/python/streaming/direct_kafka_wordcount.py]: If broker is the MUST-HAVE argument, I'd like to put it into the arguments list (before topics).

Comment (jerryshao) [inline examples/src/main/python/streaming/direct_kafka_wordcount.py]: Hi @davies , thanks for your comment, I will add this as a argument.

**Comment by jerryshao** [comment]
Hi @tdas , do we need to add a Python  version of `createRDD` for direct Kafka stream? Seems this API requires Python wrapper of Java object like `OffsetRange`.

**Comment by tdas** [comment]
Yes, python version of createRDD would be great.

BTW, is it possible to mark these experimental in Python @davies? The
Scala, Java AP is experimental as of now.

TD

**Comment by davies** [comment]
We can mark it as experimental by
```
  ::note: experimental
```

**Comment by jerryshao** [comment]
Hi @davies , should I use `::note: experimental`  or `.. note:: Experimental`, seems the MLlib codes just use the latter #3951 .

**Comment by davies** [comment]
@jerryshao please follow the examples in MLlib, thanks.

**Comment by jerryshao** [comment]
Add a new API of `createRDD`, @davies & @tdas , please help to review. Thanks a lot.

File: python/pyspark/streaming/kafka.py [hunk 2]
```diff
@@ -81,3 +79,128 @@ def getClassByName(name):
...
+        :param fromOffset: Inclusive starting offset.
+        :param untilOffset: Exclusive ending offset.
+        """
+        self._topic = topic
+        self._partition = partition
+        self._fromOffset = fromOffset
+        self._untilOffset = untilOffset
+
+    def _joffsetRange(self, sc):
+        java_import(sc._jvm, "org.apache.spark.streaming.kafka.OffsetRange")
```
Comment (davies) [inline python/pyspark/streaming/kafka.py]: It's imported by createRDD(), no more needed here.

File: python/pyspark/streaming/kafka.py [hunk 3]
```diff
@@ -81,3 +79,128 @@ def getClassByName(name):
...
+            jrdd = sc._jvm.KafkaUtils.createRDD(sc._jsc, array, array, decoder, decoder,
+                                                jparam, joffsetRanges)
+        except Py4JError, e:
+            # TODO: use --jar once it also work on driver
+            if not e.message or 'call a package' in e.message:
+                print "No kafka package, please put the assembly jar into classpath:"
+                print " $ bin/spark-submit --driver-class-path external/kafka-assembly/target/" + \
+                      "scala-*/spark-streaming-kafka-assembly-*.jar"
+            raise e
+        ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
```
Comment (davies) [inline python/pyspark/streaming/kafka.py]: There are lots of duplicated code here, could you figure out a way to re-use them? (it's OK to go as this, if it's not easy)

File: python/pyspark/streaming/kafka.py [hunk 4]
```diff
@@ -81,3 +79,128 @@ def getClassByName(name):
         ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
         stream = DStream(jstream, ssc, ser)
         return stream.map(lambda (k, v): (keyDecoder(k), valueDecoder(v)))
+
+    @staticmethod
+    def createDirectStream(ssc, brokerList, topics, kafkaParams={},
+                           keyDecoder=utf8_decoder, valueDecoder=utf8_decoder):
+        """
+        ..note:: experimental
```
Comment (davies) [inline python/pyspark/streaming/kafka.py]: Sorry, it should be 
```
.. note:: Experimental
```

**Comment by jerryshao** [comment]
Thanks @davies for your review, I will fix these comments :).

**Comment by jerryshao** [comment]
Hi @davies , mind taking a look at this again, I've addressed your comments, though some duplications are hard to remove, any suggestions?

**Comment by davies** [comment]
LGTM, thanks!

File: python/pyspark/streaming/kafka.py [hunk 5]
```diff
@@ -70,7 +71,103 @@ def createStream(ssc, zkQuorum, groupId, topics, kafkaParams={},
...
             if 'ClassNotFoundException' in str(e.java_exception):
-                print """
+                KafkaUtils._printErrorMsg(ssc.sparkContext)
+            raise e
+        ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
+        stream = DStream(jstream, ssc, ser)
+        return stream.map(lambda (k, v): (keyDecoder(k), valueDecoder(v)))
+
+    @staticmethod
+    def createDirectStream(ssc, brokerList, topics, kafkaParams={},
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: We should also allow creating direct stream with offsets and specify optional leaders. That is, all the stuff the advanced version of createDirectStream supports. The only thing that we cannot easily support in python, is the custom function MessagAndMetadata => R. Other than that we should be able to do others.

File: python/pyspark/streaming/kafka.py [hunk 6]
```diff
@@ -70,7 +71,103 @@ def createStream(ssc, zkQuorum, groupId, topics, kafkaParams={},
...
+                           keyDecoder=utf8_decoder, valueDecoder=utf8_decoder):
+        """
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: Also, taking a broker list is different from the Scala / Java API. We should keep this consistent with the Scala / Java API.

**Comment by tdas** [comment]
This is looking good, but unfortunately its hard to say due to the lack of unit tests. We have to test the kafka python API through python unit tests. I can open a separate JIRA for that, and it will be good to have that framework set up first.

**Comment by jerryshao** [comment]
Thanks @tdas for your review, maybe we should figure out a way to test the Kafka Python API at first.

**Comment by mattaylor** [comment]
Can this be merged now the API tests are done?

**Comment by tdas** [comment]
Yes, only after unit tests cover this API. @jerryshao can you add unit tests to this.

**Comment by jerryshao** [comment]
Yes, will do. I'm now working on this to add more API related to DirectKafkaInputStream as well as unit test.

File: python/pyspark/streaming/kafka.py (hunk 5 above)
Comment (jerryshao) [inline python/pyspark/streaming/kafka.py]: Hi @tdas , these days I'm trying to enable custom function `MessageAndMetadata => R` in Python related API, I just tried several solutions but unfortunately no success, do you have any suggestion?

Comment (tdas) [inline python/pyspark/streaming/kafka.py]: That is quite tricky. And its not even clear to me what the demand will be for this from the Python API. What I wanted to have in the Python aPI is the ability to specify starting offsets.

Comment (jerryshao) [inline python/pyspark/streaming/kafka.py]: Currently I have a solution, serialize the content of `MessageAndMetadata` into byte array in Scala code, and unpack, reconstruct the python object of `MessageAndMetadata` in python code. So the handler can be supported. What do you think of this solution?

Comment (tdas) [inline python/pyspark/streaming/kafka.py]: Lets leave that for another PR. That sounds complicated, and needs careful
thought.

Comment (jerryshao) [inline python/pyspark/streaming/kafka.py]: OK, I will leave this to another PR. 

Basically this solution is not so complicated, just write (topic, partition, key, message, offset) with a controlled format into a byte array. And the python code unpack this byte array with same format into the original data and construct a object used for message handler function.

Comment (tdas) [inline python/pyspark/streaming/kafka.py]: I agree as such it is not a complicated thing, just that this way of doing
things is non-standard practice (unless you show me some examples of this
being done in other python kafka libraries) and therefore should be added
only if there is a good use case / demand for it.

Comment (jerryshao) [inline python/pyspark/streaming/kafka.py]: OK, that make sense, I will ignore this handler support in Python API unless someone has specific requirement for this.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by jerryshao** [comment]
@tdas , I've added more APIs with unit tests, would you please take a look at it, thanks a lot.

**Comment by mattaylor** [comment]
any more news on this one?

**Comment by jerryshao** [comment]
Just waiting for review :).

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 7]
```diff
@@ -558,4 +560,100 @@ private class KafkaUtilsPythonHelper {
...
+      classOf[DefaultDecoder],
+      kafkaParams,
+      topics)
+  }
+
+  def createDirectStream(
+      jssc: JavaStreamingContext,
+      kafkaParams: JMap[String, String],
+      fromOffsets: JMap[TopicAndPartition, JLong])
+    : JavaPairInputDStream[Array[Byte], Array[Byte]] = {
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: I think we generally dont put ':' in the next line. Either 
```
     fromOffsets: JMap[TopicAndPartition, JLong]) : JavaPairInputDStream[Array[Byte], Array[Byte]] = 
```
or 
```
     fromOffsets: JMap[TopicAndPartition, JLong]
   ) : JavaPairInputDStream[Array[Byte], Array[Byte]] = 
```

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 8]
```diff
@@ -558,4 +560,100 @@ private class KafkaUtilsPythonHelper {
...
+        classOf[DefaultDecoder],
+        classOf[(Array[Byte], Array[Byte])],
+        kafkaParams,
+        fromOffsets,
+        messageHandler)
+    new JavaPairInputDStream(jstream.inputDStream)
+  }
+
+  def createOffsetRange(topic: String, partition: Int, fromOffset: Long, untilOffset: Long)
+      : OffsetRange = OffsetRange.create(topic, partition, fromOffset, untilOffset)
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: Same comment as above

File: python/pyspark/streaming/kafka.py [hunk 9]
```diff
@@ -70,7 +71,195 @@ def createStream(ssc, zkQuorum, groupId, topics, kafkaParams={},
...
+            if 'ClassNotFoundException' in str(e.java_exception):
+                KafkaUtils._printErrorMsg(ssc.sparkContext)
+            raise e
+
+        ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
+        stream = DStream(jstream, ssc, ser)
+        return stream.map(lambda (k, v): (keyDecoder(k), valueDecoder(v)))
+
+    @staticmethod
+    def createDirectStreamFromOffset(ssc, kafkaParams, fromOffsets,
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: I thought about this a little bit. But I think we should follow the precedent set by the `createStream` and other Python API where there is only method, with many optional parameters. So instead of having `createDirectStream` and `createDirectStreamFromOffsets`, lets just have `createDirectStream` with another optional parameter `fromOffsets`. `fromOffsets` should have the same keys as in topics, otherwise throw an error.

File: python/pyspark/streaming/kafka.py [hunk 10]
```diff
@@ -70,7 +71,195 @@ def createStream(ssc, zkQuorum, groupId, topics, kafkaParams={},
...
+            if 'ClassNotFoundException' in str(e.java_exception):
+                KafkaUtils._printErrorMsg(sc)
+            raise e
+
+        ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
+        rdd = RDD(jrdd, sc, ser)
+        return rdd.map(lambda (k, v): (keyDecoder(k), valueDecoder(v)))
+
+    @staticmethod
+    def createRDDWithLeaders(sc, kafkaParams, offsetRanges, leaders,
```
Comment (tdas) [inline python/pyspark/streaming/kafka.py]: Similar comment to createDirectStream, lets just have only one `createRDD` with `leader` as optional parameters

**Comment by tdas** [comment]
Looks almost good, except the comments on the API. Other than that, i took a detailed pass on everything else and it looks good.

File: python/pyspark/streaming/kafka.py (hunk 9 above)
Comment (jerryshao) [inline python/pyspark/streaming/kafka.py]: Since python do not support method overload, so I use different method name to differentiate it. I will change to way you mentioned.

**Comment by jerryshao** [comment]
Hi @davies and @tdas , I met a problem of converting Python `int` into Java `Long`, the Java API in KafkaUtils requires offset as `Long` type, this is simple for Python 2, since Python 2 has a built-in `long` type which can be mapped to Java `Long` through py4j automatically, but python 3 only has `int` type, and py4j will map python `int` into Java `Integer`, I'm not sure how to support `Long` in python 3.

A simple solution is to modify all the Java-Python interface to change to type `Interger`, but it may not support super large offset. I'm not sure is there any other solution. Sorry for dumb question and thanks a lot in advance.

**Comment by tdas** [comment]
@JoshRosen Any thoughts on this.

**Comment by tdas** [comment]
Offline conversation with @JoshRosen 
If it works for python 2, then make the corresponding unit test such that it will be skipped for python 3. Take a look at https://github.com/apache/spark/blob/master/python/pyspark/tests.py#L944

**Comment by jerryshao** [comment]
Thanks a lot @tdas for your comments, but what if users want to use this API in Python 3, they will still meet such problem, we only neglect the problem by skipping unit test, but not actually solve this.

**Comment by tdas** [comment]
Aaah, I forgot to mention. Please open up a JIRA regarding this pySpark
problem. Its most likely a Py4J problem, and hence needs to be followed up
with Py4J. And the python 3 is still an experimental so its okay for now.
There are other unit tests in Spark that are skipped for Python 3.

**Comment by jerryshao** [comment]
OK, thanks a lot for your suggestion, I will open a JIRA about this issue.

**Comment by tdas** [comment]
LGTM. Will merge when tests pass.

**Comment by tdas** [comment]
Alright! Merging it. Thanks! :)

**Comment by Arttii** [comment]
Hi 

I seem to be having a problem with this. The default py4j dict and set conversion does not seem to be working on my machine(Windows 8.1 Anaconda Python 2.7 64 Bit Java 1.7). I had to hack kafka.py a bit and include the old conversion code in there for this to work? Any idea what the problem might be?

**Comment by jerryshao** [comment]
Hi @Arttii , maybe it is the problem of your local environment, my machine is Ubuntu 14.04 with Python 2.7.6, it looks OK on my side, also there's no problem in Jenkins environment, where Python version is 2.6.

**Comment by jerryshao** [comment]
Would you please paste your stack trace?

**Comment by Arttii** [comment]
This is the normal WordCount example running with local[*]
```
15/04/29 10:53:48 INFO SparkContext: Running Spark version 1.3.1
15/04/29 10:53:49 INFO SecurityManager: Changing view acls to: a.topchyan
15/04/29 10:53:49 INFO SecurityManager: Changing modify acls to: a.topchyan
15/04/29 10:53:49 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users with view permissions: Set(a.topchyan); users with mod
ify permissions: Set(a.topchyan)
15/04/29 10:53:49 INFO Slf4jLogger: Slf4jLogger started
15/04/29 10:53:49 INFO Remoting: Starting remoting
15/04/29 10:53:49 INFO Remoting: Remoting started; listening on addresses :[akka.tcp://sparkDriver@DELLHR57L12.replynet.prv:52706]
15/04/29 10:53:49 INFO Utils: Successfully started service 'sparkDriver' on port 52706.
15/04/29 10:53:49 INFO SparkEnv: Registering MapOutputTracker
15/04/29 10:53:49 INFO SparkEnv: Registering BlockManagerMaster
15/04/29 10:53:49 INFO DiskBlockManager: Created local directory at C:\Users\ATOPCH~1.REP\AppData\Local\Temp\spark-af5177e7-d148-4013-a097-113c8f92ed63\blockmgr
-c45e19f6-a825-4763-a94e-eea19dbb1f26
15/04/29 10:53:49 INFO MemoryStore: MemoryStore started with capacity 265.4 MB
15/04/29 10:53:49 INFO HttpFileServer: HTTP File server directory is C:\Users\ATOPCH~1.REP\AppData\Local\Temp\spark-8474fc1d-7c66-4b25-b304-2fd743061967\httpd-e
ba80b64-1a5a-49c7-9bd2-6f7a4be543d7
15/04/29 10:53:49 INFO HttpServer: Starting HTTP Server
15/04/29 10:53:49 INFO Server: jetty-8.y.z-SNAPSHOT
15/04/29 10:53:49 INFO AbstractConnector: Started SocketConnector@0.0.0.0:52707
15/04/29 10:53:49 INFO Utils: Successfully started service 'HTTP file server' on port 52707.
15/04/29 10:53:49 INFO SparkEnv: Registering OutputCommitCoordinator
15/04/29 10:53:50 INFO Server: jetty-8.y.z-SNAPSHOT
15/04/29 10:53:50 INFO AbstractConnector: Started SelectChannelConnector@0.0.0.0:4040
15/04/29 10:53:50 INFO Utils: Successfully started service 'SparkUI' on port 4040.
15/04/29 10:53:50 INFO SparkUI: Started SparkUI at http://DELLHR57L12.replynet.prv:4040
15/04/29 10:53:50 INFO Executor: Starting executor ID <driver> on host localhost
15/04/29 10:53:50 INFO AkkaUtils: Connecting to HeartbeatReceiver: akka.tcp://sparkDriver@DELLHR57L12.replynet.prv:52706/user/HeartbeatReceiver
15/04/29 10:53:50 INFO NettyBlockTransferService: Server created on 52745
15/04/29 10:53:50 INFO BlockManagerMaster: Trying to register BlockManager
15/04/29 10:53:50 INFO BlockManagerMasterActor: Registering block manager localhost:52745 with 265.4 MB RAM, BlockManagerId(<driver>, localhost, 52745)
15/04/29 10:53:50 INFO BlockManagerMaster: Registered BlockManager
Traceback (most recent call last):
  File "D:\workingdir\app\driver.py", line 33, in <module>
    kvs = KafkaUtils.createDirectStream(ssc, ["testopic"], {"metadata.broker.list": "obiwan:9092,r2d2:9092,vader:9092"})
  File "C:/spark\python\pyspark\streaming\kafka.py", line 126, in createDirectStream
    jstream = helper.createDirectStream(ssc._jssc, kafkaParams, jtopics, jfromOffsets)
  File "C:\spark\python\lib\py4j-0.8.2.1-src.zip\py4j\java_gateway.py", line 529, in __call__
    [get_command_part(arg, self.pool) for arg in new_args])
  File "C:\spark\python\lib\py4j-0.8.2.1-src.zip\py4j\protocol.py", line 265, in get_command_part
    command_part = REFERENCE_TYPE + parameter._get_object_id()
AttributeError: 'dict' object has no attribute '_get_object_id'
Press any key to continue . . .
```
If I pass in 
```python
    jparam = MapConverter().convert(kafkaParams, ssc.sparkContext._gateway._gateway_client)
```
seems to work fine.

**Comment by Arttii** [comment]
I dont think this is a bug, but rather a problem on my side. I am just wondering what parts can impact this. There's nothing in KafkaUtils or the PythonHelper class that handles any of the conversion logic( I think), so this is a py4j thing. Well I think so at least

**Comment by jerryshao** [comment]
So can you try receiver-based Kafka stream using `createStream`? If this is a problem of py4j, I assume this will also be failed. Originally we actually use `MapConverter` to convert python dict into Java Map, but now seems this conversion is done implicitly.

**Comment by Arttii** [comment]
I did try it. Also fails. Py4J docs say that dict and set are converted to HashMap and HasSet by default. This does not seem to be happening, which is really weird. So I just put in the direct conversion and everything works. I might try py4j forums or something. This is fairly weird.

**Comment by jerryshao** [comment]
Might be the problem of py4j, I'm not the expert of py4j, what is your insight @davies ?

Maybe you could report a bug in JIRA, so others can take a crack of this problem.

**Comment by tdas** [comment]
I believe @davies will be slow in responding as he is on a vacation.
Pinging @joshrosen instead.
